runtime/lua/vim/shared.lua

```diff
@@ -1192,29 +1221,60 @@ function vim._with(context, f)
     if not vim.api.nvim_win_is_valid(context.win) then
       error('Invalid window id: ' .. context.win)
     end
+    -- TODO: Maybe allow it?
+    if context.buf and vim.api.nvim_win_get_buf(context.win) ~= context.buf then
+      error('Can not set both `buf` and `win` context.')
+    end
   end
 
-  -- Store original options
-  local previous_options ---@type table<string, any>
-  if context.options then
-    previous_options = {}
-    for k, v in pairs(context.options) do
-      previous_options[k] =
-        vim.api.nvim_get_option_value(k, { win = context.win, buf = context.buf })
-      vim.api.nvim_set_option_value(k, v, { win = context.win, buf = context.buf })
+  -- Decorate so that save-set-restore options is done in correct window-buffer
+  local callback = function()
+    -- Save and set context (NOTE: triggers `OptionSet` event)
+    local cache = { bo = {}, env = {}, go = {}, wo = {} } ---@type table<string, any>
+
+    -- Use specific order from possibly most to least intrusive
+    for _, scope in ipairs(scope_order) do
+      for name, context_value in pairs(context[scope] or {}) do
+        local sc = scope == 'o' and scope_map[vim.api.nvim_get_option_info2(name, {}).scope]
+          or scope
+
+        -- Do not override already set cache and fall back to `vim.NIL` to
+        -- cache `nil` values (which needs restoring later)
+        cache[sc][name] = cache[sc][name] or vim[sc][name] or vim.NIL
```

Comment:
Do I understand it correctly that e.g. if both `o` and `bo` scopes are used with different values then `o` will be used since it has higher priority?

Reply:
On the contrary, `bo` will be used as it will be set *after* setting the `o` context.

Using `cache[sc][name] = cache[sc][name] or vim[sc][name]` instead of direct `cache[sc][name] = vim[sc][name]` guards exactly from this kind of case. Without it, cache would have been set to user's value from the `o` context and later (erroneously) restored to it.

Reply:
But good question, as I realized that there was no test case specifically if same option is supplied in several contexts.